feat: rewrite audit script in Python

HastD opened the pull request:
The new Python script currently carries out the same checks as the old bash audit script (with a few minor enhancements, like checking for `libhardened_malloc-light.so` in a couple places), plus one additional check for MAC randomization. It is more modular and significantly faster: on my system with lots of flatpaks installed, the bash script typically takes about 20 seconds to run, while the Python script takes about 9 seconds.

Flatpak permission checks can be skipped by passing the `--skip flatpak` option. Note: the bash script seems to have only checked one version of each flatpak in cases where multiple versions are installed—I think this was an error, though in practice it typically just affects runtimes, which aren't visible in Flatseal anyway and are unlikely to fail the permission checks.

Implementation details: Each check is a generator (optionally asynchronous) that optionally accepts some state (an associative array that can be used to pass information between checks) and yields a sequence of reports (the lines that are shown under the "Audit" heading with a "success/warning/failure" status) and recommendations (printed at the end), which are handled by an audit runner. A decorator `@audit` is used to queue up generators to be run by the default audit runner. The file can also be imported as a Python module for use in other code—it doesn't run any audits automatically if imported.

HastD committed:
feat: rewrite audit script in Python.

The new Python script is a standalone executable in
/usr/bin/audit-secureblue. It currently carries out the same checks as
the old bash audit script, but is faster and more modular.

Each check is a generator that accepts some state (an associative array
that can be used to pass information between checks) and yields a
sequence of reports and recommendations, which are handled by the audit
runner.

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
feat: support async audits, make flatpak checks concurrent.

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

fiftydinar commented:
Maybe it's smart to put `audit-secureblue` to `/usr/libexec/secureblue/audit-secureblue`?

This way, path needs to be called fully for binary to execute & users will rely on `ujust` recipe more than just directly calling `audit-secureblue`?

Also, this way, any `secureblue` related scripts can be put into `/usr/libexec/secureblue/`, which would look neat.

HastD committed:
fix: don't error out if secureboot is unsupported

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
move Python audit script to libexec, set up to call via ujust

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

Rubiginosa commented:
I think the flatpak order could still be made deterministic—and that it might be worth the effort, especially running repeated tests or comparing across installations.

Not familiar with the specifics in python, but having the output thread wait on each of the checking threads in order should work.

HastD committed:
yield flatpak permission checks in lexicographic order

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD commented:
Thanks for the feedback, I put the script in `/usr/libexec/secureblue/audit_secureblue.py` (with `ujust audit-secureblue` simply calling this and forwarding any passed arguments) and made the flatpak permission checks print out in lexicographic order.

I see that Codacy is flagging the use of the `subprocess` module in the script; however, every use of it here is calling commands with trusted inputs and with `shell=False`, so there shouldn't be any security concern. Do I need to add some marker to the script to tell Codacy to ignore it, or can that check just be overridden some other way?

HastD committed:
rename file to follow Python convention

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
many minor changes recommended by linter + formatter

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
change a couple names for clarity

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
split flatpak remotes audit into separate function

Also simplify implementation of async flatpak permissions checks.

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
suppress bandit security warnings for subprocess

The security issue is only a concern with untrusted inputs, and we only
call subprocess with trusted inputs.

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD commented:
Hm, just noticed the error handling with this script needs a bit of work, not great that it spits out an irrelevant stack trace if the user hits Ctrl+C... one sec, I'm going to add an interrupt handler to deal with that better.

HastD committed:
add interrupt handler to avoid dumping stack trace on Ctrl+C

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
remove unnecessary 'async'

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
improve sigint handling; allow skipping kargs checks

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
Added check for MAC randomization.

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
allow stateless checks to be defined with no arguments

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -0,0 +1,986 @@
+                color_code = 33  # yellow
+            case Status.FAILURE:
+                color_code = 31  # red
+        return f"\x1b\x5b{color_code}m{self.name}\x1b\x5b39m"
```
if there's some way to make this more readable, that would be nice :sweat_smile:

HastD replied:
ANSI escape sequences aren't the most readable... I can at least write `\x5b` as `[`, I only did that in the first place because it was making my editor confused about indentation (I need to get a Python LSP set up properly in Vim...) and I forgot to change it back.

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -0,0 +1,986 @@
+    WARNING = enum.auto()
+    FAILURE = enum.auto()
+
+    def in_color(self) -> str:
```
`to_string_in_color`?

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -0,0 +1,986 @@
+            self.warnings = warnings
+
+    def __str__(self) -> str:
+        s = f"{self.description + '...':<68} [ {self.status.in_color()} ]"
```
better variable name like `status_string_in_color` maybe?

HastD replied:
I'll call it `report_str` since it includes not just the status but also the subsequent warnings.

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -0,0 +1,986 @@
+        self,
+        name: str,
+        call: AsyncGenerator[Any, [dict[str, Any]]],
+        *,
```
what is this?

HastD replied:
The `*` marks all subsequent function arguments as keyword-only arguments. This helps to avoid the readability issues caused by too many positional arguments.

RoyalOughtness replied:
@HastD oh wow, that is very strange syntax

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -0,0 +1,986 @@
+    def __init__(
+        self,
+        name: str,
+        call: AsyncGenerator[Any, [dict[str, Any]]],
```
`callback` maybe?

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -0,0 +1,986 @@
+        else:
+            gen = (self.call)()
+        async for result in gen:
+            if isinstance(result, tuple):
```
should we standardize the result type instead of using Any? seems a bit messy to have to parse the type at runtime

HastD replied:
Good idea, I'll include the recommendations in the `Report` object so the generators always yield a `Report` (which has a list of recommendations that can be empty).

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -0,0 +1,986 @@
+    if os.path.isfile("/etc/sway/config.d/99-noxwayland.conf"):
+        status = SUCCESS
+    else:
+        status = FAILURE
```
maybe recommend the ujust here?

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -0,0 +1,986 @@
+    if allowed == "false":
+        status = SUCCESS
+    else:
+        status = FAILURE
```
rec here too?

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -0,0 +1,986 @@
+                if key == "ghns" and value == "false":
+                    status = SUCCESS
+    except FileNotFoundError:
+        return
```
and here

HastD replied:
Added a recommendation, let me know if the wording needs changing.

RoyalOughtness replied:
@HastD they can just run `ujust toggle-ghns`

HastD replied:
Oops, didn't notice that was a thing. Fixed.

RoyalOughtness commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -0,0 +1,986 @@
+
+
+###############################################################################
+# Checks to be run go above this line.
```
could be cleaner to move the class definitions for Report, Check, etc to their own files for organizational reasons

HastD replied:
Done, split off the abstract auditing framework to an `auditor` module. I left a few helper functions in the main `audit_secureblue.py` file since they seem to conceptually fit better there. Does this look like a reasonable way to organize it now?

ShadowSlayer1441 commented on files/system/usr/libexec/secureblue/audit_secureblue.py:
```diff
@@ -0,0 +1,986 @@
+        if f.read().strip() == "3":
+            status = SUCCESS
+            state["ptrace_allowed"] = False
+        else:
```
Maybe have a WARNING option if it is set to 1 to enable anticheat/GDB etc

HastD committed:
Merge branch 'live' into python-audit

HastD committed:
Added more recommendations; minor refactoring for clarity.

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
split abstract auditing framework into separate module

Signed-off-by: Daniel Hast <hast.daniel@protonmail.com>

HastD committed:
fix glob handling in sysctl check

HastD committed:
guard against accidental reassignment of statuses with Final

HastD committed:
fix KDE GHSN recommendation; remove unnecessary .strip()

HastD committed:
fix name of bash_completion file (see also #1025)
